boot/setefibootvars.go

```diff
@@ -0,0 +1,178 @@
+			return 0xFFFF, nil, err
+		}
+		varNumber := uint16(varNumberUint64)
+		// Read the current contents of the variable
+		varData, varAttrs, err := efiReadVariable(varName, varGUID)
+		if err != nil {
+			// Should not occur, unless variable changed since ListVariables()
+			return 0xFFFF, nil, err
+		}
+		// Parse the variable data into a efi.LoadOption so it can be used
+		dataReader := bytes.NewReader(varData)
+		varLoadOption, err := efi.ReadLoadOption(dataReader)
+		if err != nil {
+			return 0xFFFF, nil, err
+		}
+		varDevicePathStr := varLoadOption.FilePath.String()
+		if varDevicePathStr == devicePath {
+			// existing variable points to correct device path, use it
+			bootNum = varNumber
+			bootNumVarAttrs = varAttrs
+		}
+		usedBootNums[varNumber] = true
+	}
+	if bootNum == 0xFFFF {
+		// No match, find first unused boot number
+		for bootNum = 0; bootNum < 0xFFFF; bootNum += 1 {
+			if usedBootNums[bootNum] == false {
+				break
+			}
+		}
+	}
+	// XXX: there is a (miniscule) possibility that all boot numbers are used
+	varName := fmt.Sprintf("Boot%04X", bootNum)
+	err = efiWriteVariable(varName, efi.GlobalVariable, bootNumVarAttrs, data)
+	return bootNum, usedBootNums, err
+}
+
+// Reads the current BootOrder variable, inserts the given firstBootNum at the
+// beginning of the number list (and removes it from elsewhere in the list if
+// it occurs), removes any unused boot numbers from the list, and writes the
+// list as the new BootOrder variable.
+func setBootOrderVariable(firstBootNum uint16, usedBootNums map[uint16]bool) error {
+	origData, attrs, err := efiReadVariable("BootOrder", efi.GlobalVariable)
+	if err == efi.ErrVarNotExist {
+		attrs = efi.AttributeNonVolatile | efi.AttributeBootserviceAccess | efi.AttributeRuntimeAccess
+		origData = make([]byte, 0)
+	} else if err != nil {
+		return err
+	}
+	bootOrder := make([]byte, len(origData)+2)
+	binary.LittleEndian.PutUint16(bootOrder, firstBootNum)
+	i := 2
+	for len(origData) >= 2 {
+		next := binary.LittleEndian.Uint16(origData)
+		origData = origData[2:]
+		if next != firstBootNum && usedBootNums[next] == true {
+			binary.LittleEndian.PutUint16(bootOrder[i:], next)
+			i += 2
+		}
+	}
+	// If some boot numbers in BootOrder have a Boot#### which no longer
+	// exists, then they were not included in bootOrder, so truncate it to its
+	// actually used size.
+	bootOrder = bootOrder[:i]
+	return efi.WriteVariable("BootOrder", efi.GlobalVariable, attrs, bootOrder)
+}
+
+// SetEfiBootVariables sets the Boot#### and BootOrder variables for the given path.
+//
+// Writes the EFI boot variable Boot#### to contain the device path of the given
+// seedShimPath. Then, sets the BootOrder variable so that the #### number from
+// the chosen Boot#### is first, and leaves other numbers in the boot order so
+// long as there are still Boot#### variables with the corresponding number.
+func SetEfiBootVariables(seedShimPath string) error {
```

Comment:
I am nitpicking here, but there are many references to shim in this file that feel unnecesary as the code could be used with any EFI binary (seedShimPath could be efiFilePath, etc.)

Reply:
Great point, thanks!